### DISCUSSION TITLE
Cruiser open source

### DISCUSSION THREAD

**Discussion by devemux86** [devemux86/cruiser#509] [General]
I'm thinking if it's worth it to make Cruiser open source.

Is there any interest from the community and other developers?

More users can then contribute new features and improvements.

**Comment by DIrgi** [comment]
As I wrote in another tread where it originally popped up, I see myself very limited to dig deeper into a programming support.
I even think about how and when I could start participating in the Cruiser WIKI and place there some contributions where I was al least one of the guys initiating this request.

**Comment by bios4** [comment]
For me it is enough to do application coding during my daily job. 😞

Additionally, I have never written code for mobile apps and am currently not interested in getting into the subject.
I like to "contribute" in the way I'm currently doing here in Cruiser Discussions (and am still also interested in helping building up the Wiki!), but I'm not into getting a real Cruiser app coder/programmer at the moment...

**Comment by MotoUKRider** [comment]
I thought I had commented on this, but obviously not!

Having looked at the OSMand project, I find myself kind of overwhelmed with GIT this, PULL that, MANIFEST those etc.  I'm none the wiser as to exactly *what* development environment I should use.  I consider myself a reasonably fast learner but I find the prospect of staring at XML to "draw" a button on the screen rather unappealing!

There's an awful lot of scripting involved, and it's not for the faint-hearted (my recent efforts at a map theme for both Cruiser and BaseCamp have put the coffee maker through its paces).  I looked at OpenMapTiles which had potential but it wasn't immediately obvious how I would connect a routing engine to it (I assume lots of Lat/Lon calls?).

I can picture the kind of app I'd write, I can probably draw some 8-bit mock-ups in MSPaint, but actually generating the GUI components, not even close!  I like the idea of being able to open the project, make a relatively simple change and upload it for others to enjoy.

If anything, it makes you understand how much work goes into learning a language, debugging a new process and dealing with users' comments like: "can you move it to the left a bit?"... It's more likely that a developer would get into motorcycles than a rider wanting to learn how to code - the FB touring groups suggest that most prefer a paper list shoved in the tank bag :scroll: 

Android Studio, Visual Studio, Flutter, something else?  I don't want to find myself in motorcycle terms: trying to remove a wheel nut with a pair of tweezers, so what would somebody need on their (Windoze) computer to work on Cruiser?

**Reply by devemux86** [reply]
@MotoUKRider thanks for the comment. :+1: 

Only need Android and IDEA or Android Studio and open the project. All work automatically.
Similar implementation to the open source map libraries I'm developing: [Mapsforge](https://github.com/mapsforge/mapsforge) + [VTM](https://github.com/mapsforge/vtm).

In addition to what you describe, developers have to deal with each new version of Android,
that changes (often greatly) how they work. :slightly_smiling_face:

**Reply by MotoUKRider** [reply]
Well, that's encouraging :grinning:

I've done a lot with Arduino using C++ so I'm not totally against coding!

**Reply by MubarakNative** [reply]
@MotoUKRider I’m a little curious—who was the original developer of Cruiser? 🤔 I’m kinda confused about its history.

**Reply by devemux86** [reply]
@MubarakNative 

@devemux86 is the author of **[Cruiser](https://github.com/devemux86/cruiser)** application and developer of **[Mapsforge](https://github.com/mapsforge)** map library.

@MotoUKRider is the author of **[Motorider](https://github.com/devemux86/cruiser/discussions/612)** map theme.

**Reply by MubarakNative** [reply]
Thanks for your reply!

**Comment by F3llFr0mTh3Sky** [comment]
Well, there is no GPS navigation program for Linux except for NavIt, and the last time I tried it, it was just completely broken since it just kept crashing. I'm not even talking about other softwares like Gnome Maps or KDE Marble since to my knowledge they don't even support GPS navigation and might not even supports downloading offline maps. 

So open sourcing at least the desktop part of Cruiser would be very nice.
If Linux support is missing, I'd be happy to work on it. @devemux86

**Reply by devemux86** [reply]
@F3llFr0mTh3Sky

Desktop Cruiser runs on all desktop platforms: Windows, macOS and Linux:
https://wiki.openstreetmap.org/wiki/Cruiser#Installation

Arch Linux users can also install Cruiser from the [AUR](https://aur.archlinux.org/packages/cruiser/).

**Reply by F3llFr0mTh3Sky** [reply]
Oh nice. I didn't notice the desktop app was written in Java.

hmm ... does it supports maps from https://planet.openstreetmap.org/?
They're downloadable as torrent as well (just interface aria2x).

**Reply by devemux86** [reply]
@F3llFr0mTh3Sky Offlne Mapsforge maps and some online maps.

**Comment by ewaldc** [comment]
I would be willing to contribute and skilled in Java (actually mostly Kotlin), Android development (compose, sql etc.) and Android Studio.  I have a bit of knowledge in Mapsforge/VTM, but it's rather limited to playing around with the demo's and development of a render theme.  I did start to build a Tile server (based on VTM), but Telemaxx was ahead of me as he (smartly) leveraged code from Mobac. That said, I would not directly know what to to contribute to Cruiser, it is rather complete (and also perfect to test my theme).  A few areas where I might be able to contribute:

- it is not ideal that on Android you can't access the offline maps from a common folder that is e.g shared with another GPS app such as OruxMaps.  But I know that these things are made very hard by Android 11+.  But that is an area where I might be able to help...  
- some small improvements could be the addition of a (initial) route (re)calculation function e.g. when a route can not be found (e.g. due to router service or selection of wrong profile/travel mode such as using car instead of bike).  Currently I have to move the points slightly for the route to be (re)calculated.  Once the initial route present, there is the handy "recalculate function".

Other than that it really depends on what the target use for the app would be in the future and what the target use cases are.  For actual use during hiking, there are apss that are more functional such as Locus, Orux etc. but for e.g. exploring and route creation, Cruiser is far better and faster.  QMapShack etc. are more focused on this part, but you need some extension (Mapsforge for QMapShack) to use e.g. OpenAndroMaps.  Currently I have not found a way to export a route with turn directions to e.g. Orux (maybe I missed something...).

Is this the type feedback you were looking for ?

**Reply by devemux86** [reply]
@ewaldc Thanks for the detailed answer.

_I am also the developer of Mapsforge and VTM libraries._

---

> it is not ideal that on Android you can't access the offline maps from a common folder

Common folders now require scoped storage access that is slow.
While file API always has the fastest I/O access.

You could enable the extended menu and open external map files from common folders.

For other suggestions, please let's discuss them in new threads. :slightly_smiling_face:

**Reply by ewaldc** [reply]
@devemux86, apoligies, I did not mean to hijack the discussion.  I (perhaps wrongly) assumed that the idea of "open sourcing" is tightly coupled with some vision of how "cruiser" could/should evolve and, within that context, what could (candidate) developers be able to contribute. 
With a vision for cruiser, it would become easier for to express what features add value and express what could be contributed (ideas/code) and then, hopefully, for you to evaluate if it's worth to "open source".
For example, "cruiser" is today superb for exploration of maps and route creation.  But for actual use during trips it falls short in functionality e.g. the ability to load a/the map corresponding to your current location. If becoming a hiking/biking/... companion app is part of this vision (with real time rerouting, turn-by-turn  directions etc.) than for example it would make it easier to provide feedback on what we could contribute. For example, I was thinking about a WearOS companion app for hikers/bikers that keeps you on route and reroutes if needed. 
Hopefully this makes any sense ....
  
BTW, thanks for the tip on the extended menu, it is working brilliantly!   And yes, you are very right: scoped storage is much slower.  That said, on my phone the "external map" is still extremely fast, way faster than any other android app.  I was unable though to make the "external theme" work, but I will open a new discussion for that.

**Reply by devemux86** [reply]
@ewaldc External themes requite to open their downloaded zip file.

(there is a toast message with the expected extension in open dialog)

**Comment by Sonwon1** [comment]
I will just comment on this from a users prospective. I have not had the opportunity to dig deeply into using Cruiser. I hope to begin some test rides this week. So my prospective is rather limited.

Making Cruiser open source could work however I would strongly caution how this is implemented. I will speak of my experience with Osmand. Osmand has almost everything I want in navigation software. However I will not use it. It appears to me Osmand is moving to fast with new features and never addresses the bugs. I know some of the bugs get fixed however major ones never seem to be fully resolved. For example, the routing in Osmand is still hit or miss. They have made major speed improvements to the router but the router will still route through roadside parking lots from time to time. This problem has been there for over a year. GPX file support for routes and waypoints is not great and usage is clunky. They have a very narrow focus on GPX track files. I could go on however I will stop here.

My suggestions, the primary focus should be to fix the bugs. There is nothing that will taint a product's name like being buggy. Secondary focus sould be new feature request. Third make sure the menus and UI are intuitive. Fourth, new features need to be incorporated into the wiki and How To's. Cruiser just needs to work, Kurviger does a great job with it just working. They are new feature adverse to some suggestions since it is outside the scope of their vision.

Someone who is engaged in the success of Cruiser needs to have oversight of the project. I would be looking for a Linus Torvald's type of guy that will not allow buggy code or hard to maintain code into the project. Or a three person board to review oversight maybe one person is to much power? It does work for the Linux kernel though.

I think that sums up my thoughts.

**Reply by devemux86** [reply]
> Linus Torvald's type of guy that will not allow buggy code or hard to maintain code into the project

@Sonwon1 Linus does a great job managing the Linux kernel.

**Comment by raherygino** [comment]
I'm interested in this project because I've already used Mapsforge in an Android project and I'd like the desktop version if possible; I'd be happy to contribute.
